bip-0442.md

```diff
@@ -0,0 +1,253 @@
+of potentially controversial new behaviors, such as novel 2-way peg mechanisms.
+
+Alternatively `OP_RETURN` could be used to ensure the availability of the state
+recovery data, as `OP_CHECKTEMPLATEVERIFY` naturally commits to all outputs.
+However, its cost in weight units would be over 4 times higher than that of
+using `OP_PAIRCOMMIT`.
+
+One way to think about the 3 opcodes (`OP_CHECKSIGFROMSTACK`, `OP_INTERNALKEY`,
+`OP_PAIRCOMMIT`) is we decompose a `OP_CHECKSIGFROMSTACK` variant that can use
+a 1-byte `OP_TRUE` public key (substituting for the *taproot internal key*) and
+can commit to a number of stack elements as a message.
+
+### Behaviors LNhance tries to avoid introducing
+
+The following behaviors are out of scope for LNhance and should not be enabled
+as a side effect without explicit consensus:
+
+* Fine-grained introspection
+* State-carrying covenants
+* Bigint operations
+* New arithmetic capabilities using lookup tables
+
+### Alternative approaches
+
+The following list of alternative approaches were discussed and rejected for
+various reasons, either for expanding the scope or for unnecessary complexity:
+
+* OP_CAT
+* SHA256 streaming opcodes
+* Merkle operation opcodes
+* 'Kitty' CAT: result or inputs arbitrarily limited in size
+* OP_CHECKTEMPLATEVERIFY committing to the taproot annex in tapscript
+* OP_CHECKSIGFROMSTACK on n elements as message
+* OP_VECTORCOMMIT: generalized form for n > 2 elements
+
+### Cost comparison of LN-Symmetry constructions
+
+| Method        | ChannelSc | UpdateSc | UpdateW | ForceC  | Contest | Settle |
+| :------------ | --------: | -------: | ------: | ------: | ------: | :----: |
+| APO-Annex     |      8 WU |   113 WU |  100 WU | 1221 WU |  627 WU | SigOp  |
+| APO-Return    |      8 WU |   113 WU |   66 WU | 1359 WU |  765 WU | SigOp  |
+| CTV+CSFS+IKEY |     10 WU |    48 WU |   98 WU | 1328 WU |  732 WU |  CTV   |
+| CTV+CSFS      |     43 WU |    81 WU |   98 WU | 1394 WU |  765 WU |  CTV   |
+| LNhance       |     11 WU |    49 WU |  131 WU | 1191 WU |  594 WU |  CTV   |
+
+*ChannelSc: channel script, UpdateSc: update script, UpdateW: witness is the
+same size for both Force Close and Contest in LN-Symmetry, ForceC: total cost of unilateral close transactions*
+
+### Proving general computation
+
+Merkle trees can be used to prove computation where the root of the tree
+represents the *function* and the leaves represent the *inputs* and *output*.
+There are practical limits to the entropy space for the *inputs* as they need
+to be iterated over and hashed into a merkle root.
+
+MAST trees can currently cover 128 bits of entropy space, which is well over
+the practical limits to iterate over and merklize. Therefore, we assume this
+capability does not materially extend what computations are possible to prove
+in bitcoin script. While `OP_PAIRCOMMIT` is not limited to a height of 128,
+that should not be practically feasible to utilize.
```

Comment:
should not be -> is

Reply:
My confidence in that was not a 100% at the time. It was more of an open question posed to the dev community to falsify.

Reply:
> should not be -> is

as in, it *is* practical to utilize deeper trees? (for this technique)

Did I miss a discovery?

I suppose a sparse tree could be deeper but that's logically equivalent to reducing the bit size of the inputs afaict.

Reply:
Sorry, "is not"